theano/sandbox/gpuarray/opt.py

```diff
@@ -371,52 +463,62 @@ def local_gpua_incsubtensor(node):
 
 @register_opt('fast_compile')
 @op_lifter([tensor.AdvancedIncSubtensor1])
-def local_gpua_advanced_incsubtensor(node):
+def local_gpua_advanced_incsubtensor(node, context):
+
+    x, y = node.inputs[0:2]
+    coords = node.inputs[2:]
+
+    ctx = get_context(context)
 
     # This optimization is disabled if cuda is not active
-    if pygpu.get_default_context().kind != "cuda":
+    if ctx.kind != "cuda":
         return None
 
-    x, y = node.inputs[0:2]
-    coords = node.inputs[2:]
     set_instead_of_inc = node.op.set_instead_of_inc
-    active_device_no = theano.sandbox.cuda.active_device_number()
-    device_properties = theano.sandbox.cuda.device_properties
 
-    compute_capability = device_properties(active_device_no)['major']
+    # HACK ALERT: currently the bin_id for cuda contexts is sm_<major><minor>
+    # this is not set in stone and may change later
+    cc = ctx.bin_id
+    assert cc.startswith('sm_')
+    compute_capability = int(cc[3])
 
     if (compute_capability < 2 or x.ndim != 2 or y.ndim != 2):
         return GpuAdvancedIncSubtensor1(
+            inplace=node.op.inplace,
             set_instead_of_inc=set_instead_of_inc)
     else:
         return GpuAdvancedIncSubtensor1_dev20(
-            set_instead_of_inc=set_instead_of_inc)
+            inplace=node.op.inplace,
+            set_instead_of_inc=set_instead_of_inc,
+            context=context)
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.CAReduce, tensor.Sum, tensor.elemwise.Prod])
-def local_gpua_careduce(node):
+def local_gpua_careduce(node, context):
     if isinstance(node.op.scalar_op, (scalar.Add, scalar.Mul,
                                       scalar.Maximum, scalar.Minimum)):
-        dev = theano.sandbox.gpuarray.init_dev.device
-        if dev.startswith('opencl'):
-            op = GpuCAReduceCPY
+        x, = node.inputs
+        ctx = get_context(context)
+        if ctx.kind == 'opencl':
             if node.op.scalar_op not in [scalar.add, scalar.mul]:
                 # We don't support yet all reduction with cpy code.
                 return
```

Comment:
Just to note that the return it still there. So we still disable it. But I do not want to add new stuff in this PR, so if you want to change that, in a new PR.